File: flutter/packages/workout_log_core/lib/src/analytics/cycle_report.dart
import '../coding.dart';
import '../models/block.dart';
import '../models/exercise.dart';
import '../models/session.dart';
import '../models/work_set.dart';
import 'session_metrics.dart';

/// A finding worth a training decision, tagged with the principle it comes from.
/// The system reports; the programming decisions stay the athlete's.
class TrainingAlert {
  const TrainingAlert({
    required this.principle,
    required this.message,
    this.sessions = const [],
  });

  /// `P3`, `P5`, `P8`, `P10` — see `docs/01-training-principles.md`.
  final String principle;
  final String message;
  final List<String> sessions;

  Map<String, dynamic> toJson() {
    final json = <String, dynamic>{
      'principle': principle,
      'message': message,
    };
    if (sessions.isNotEmpty) json['sessions'] = sessions;
    return json;
  }
}

/// How often a movement pattern was trained — the frequency variety is bought
/// with (P8). Below roughly one slot per cycle, linear progression is dead.
class PatternFrequency {
  const PatternFrequency({
    required this.pattern,
    required this.slots,
    required this.sessions,
    required this.perCycle,
  });

  final MovementPattern pattern;
  final int slots;
  final int sessions;
  final double perCycle;

  Map<String, dynamic> toJson() => {
        'pattern': pattern.name,
        'slots': slots,
        'sessions': sessions,
        'per_cycle': perCycle,
      };
}

/// Strength and conditioning side by side, never fused into one score: fatigue
/// is additive and the point is to see when both rise together (P7).
class LoadPoint {
  const LoadPoint({
    required this.date,
    required this.cycleDay,
    required this.kind,
    required this.tonnageKg,
    required this.sets,
    this.metconSec,
    this.workMin,
  });

  final String date;
  final String cycleDay;
  final Kind kind;
  final double tonnageKg;
  final int sets;
  final double? metconSec;
  final double? workMin;

  Map<String, dynamic> toJson() {
    final json = <String, dynamic>{
      'date': date,
      'cycle_day': cycleDay,
      'kind': kind.name,
      'tonnage_kg': tonnageKg,
      'sets': sets,
    };
    put(json, 'metcon_sec', metconSec);
    put(json, 'work_min', workMin);
    return json;
  }
}

/// What a stretch of sessions says as a block rather than one at a time: band
/// distribution, pattern frequency, combined load, and the alerts those raise.
class TrainingReport {
  TrainingReport({
    required this.from,
    required this.to,
    required this.days,
    required this.cycleLengthDays,
    required this.sessionCount,
    required this.kinds,
    required this.bandSets,
    required this.bandSlots,
    required this.patterns,
    required this.load,
    required this.timeDomains,
    required this.alerts,
    required this.unknownExercises,
  });

  /// [volumeSlotShareThreshold] guards P5, the failure that actually happened:
  /// a block with every slot in the volume range left no light session to
  /// recover against and ended in an involuntary deload. P4 rations volume work
  /// to 2-3 slots per cycle, which is where the default comes from.
  factory TrainingReport.of(
    List<Session> sessions, {
    required Catalogue catalogue,
    int cycleLengthDays = 12,
    double volumeSlotShareThreshold = 0.3,
  }) {
    final ordered = [...sessions]..sort((a, b) => a.date.compareTo(b.date));
    final bandSets = <RepBand, int>{};
    final bandSlots = <RepBand, int>{};
    final patternSlots = <MovementPattern, int>{};
    final patternSessions = <MovementPattern, Set<String>>{};
    final timeDomains = <TimeDomain, int>{};
    final kinds = <Kind, int>{};
    final load = <LoadPoint>[];
    final unknown = <String>{};
    final alerts = <TrainingAlert>[];
    final explosiveOffenders = <String>{};
    final decayOffenders = <String>{};

    for (final session in ordered) {
      final metrics = SessionMetrics.of(session);
      kinds[session.kind] = (kinds[session.kind] ?? 0) + 1;
      metrics.bandSets.forEach((band, n) => bandSets[band] = (bandSets[band] ?? 0) + n);
      metrics.bandSlots.forEach((band, n) => bandSlots[band] = (bandSlots[band] ?? 0) + n);
      for (final metcon in metrics.metcons) {
        final domain = metcon.domain;
        if (domain != null) timeDomains[domain] = (timeDomains[domain] ?? 0) + 1;
      }
      load.add(LoadPoint(
        date: session.date,
        cycleDay: session.cycleDay,
        kind: session.kind,
        tonnageKg: metrics.tonnageKg,
        sets: metrics.sets,
        metconSec: metrics.metcons.isEmpty
            ? null
            : metrics.metcons.fold<double>(
                0, (sum, m) => sum + (m.totalSec ?? 0)),
        workMin: metrics.density.workMin > 0 ? metrics.density.workMin : null,
      ));

      for (final block in session.blocks) {
        final names = switch (block) {
          StrengthBlock(:final exercise) => [exercise],
          MetconBlock() => [for (final e in block.exercises) e.name],
          _ => const <String>[],
        };
        for (final name in names) {
          final resolved = catalogue.resolve(name);
          if (resolved == null) {
            if (name.trim().isNotEmpty) unknown.add(name);
            continue;
          }
          final pattern = resolved.pattern;
          if (pattern == null) continue;
          if (block is StrengthBlock) {
            patternSlots[pattern] = (patternSlots[pattern] ?? 0) + 1;
          }
          patternSessions.putIfAbsent(pattern, () => {}).add(session.date);
        }

        if (block is StrengthBlock) {
          final resolved = catalogue.resolve(block.exercise);
          if (resolved != null && resolved.isExplosive) {
            final label = '${session.date} ${block.exercise}';
            if (block.sets.any((s) => (s.recordedReps ?? 0) > 3)) {
              explosiveOffenders.add(label);
            }
            if (_repsCollapse(block.sets)) decayOffenders.add(label);
          }
        }
      }
    }

    final from = ordered.isEmpty ? null : ordered.first.date;
    final to = ordered.isEmpty ? null : ordered.last.date;
    final days = _daysBetween(from, to);
    final cycles = days == null ? null : days / cycleLengthDays;

    final patterns = [
      for (final pattern in MovementPattern.values)
        if ((patternSlots[pattern] ?? 0) > 0 ||
            (patternSessions[pattern]?.isNotEmpty ?? false))
          PatternFrequency(
            pattern: pattern,
            slots: patternSlots[pattern] ?? 0,
            sessions: patternSessions[pattern]?.length ?? 0,
            perCycle: cycles == null || cycles <= 0
                ? 0
                : _round((patternSlots[pattern] ?? 0) / cycles),
          ),
    ];

    final totalSlots = bandSlots.values.fold<int>(0, (a, b) => a + b);
    final volumeSlots = bandSlots[RepBand.volume] ?? 0;
    if (totalSlots > 0 && volumeSlots / totalSlots > volumeSlotShareThreshold) {
      alerts.add(TrainingAlert(
        principle: 'P5',
        message: 'volume (7+) slots are '
            '${(volumeSlots / totalSlots * 100).round()}% of $totalSlots slots, '
            'over the ${(volumeSlotShareThreshold * 100).round()}% guard — the '
            'state that produced the involuntary deload',
      ));
    }
    if (explosiveOffenders.isNotEmpty) {
      alerts.add(TrainingAlert(
        principle: 'P3',
        message: 'explosive lifts logged above 3 reps — bar speed decays and the '
            'lift is trained in the wrong zone',
        sessions: explosiveOffenders.toList()..sort(),
      ));
    }
    if (decayOffenders.isNotEmpty) {
      alerts.add(TrainingAlert(
        principle: 'P3',
        message: 'reps collapse across the ramp on an explosive lift '
            '(the 4 / 2 / 1 pattern)',
        sessions: decayOffenders.toList()..sort(),
      ));
    }
    if (cycles != null && cycles >= 1) {
      final starved = [
        for (final p in patterns)
          if (p.perCycle < 1) '${p.pattern.name} (${p.perCycle}/cycle)',
      ];
      if (starved.isNotEmpty) {
        alerts.add(TrainingAlert(
          principle: 'P8',
          message: 'trained below once per $cycleLengthDays-day cycle, so linear '
              'progression on them is dead: ${starved.join(', ')}',
        ));
      }
    }
    if (cycles != null &&
        cycles >= 2 &&
        (kinds[Kind.deload] ?? 0) == 0 &&
        (kinds[Kind.retest] ?? 0) == 0) {
      alerts.add(TrainingAlert(
        principle: 'P10',
        message: 'no deload or retest in ${days!} days — deloads are hygiene, '
            'and progress between anchors is the unit of evaluation',
      ));
    }

    return TrainingReport(
      from: from,
      to: to,
      days: days,
      cycleLengthDays: cycleLengthDays,
      sessionCount: ordered.length,
      kinds: kinds,
      bandSets: bandSets,
      bandSlots: bandSlots,
      patterns: patterns,
      load: load,
      timeDomains: timeDomains,
      alerts: alerts,
      unknownExercises: unknown.toList()..sort(),
    );
  }

  final String? from;
  final String? to;
  final int? days;
  final int cycleLengthDays;
  final int sessionCount;
  final Map<Kind, int> kinds;
  final Map<RepBand, int> bandSets;
  final Map<RepBand, int> bandSlots;
  final List<PatternFrequency> patterns;
  final List<LoadPoint> load;
  final Map<TimeDomain, int> timeDomains;
  final List<TrainingAlert> alerts;

  /// Names no catalogue entry resolves: surfaced, never silently dropped (ADR-006).
  final List<String> unknownExercises;

  double get totalTonnageKg =>
      _round(load.fold<double>(0, (sum, p) => sum + p.tonnageKg));

  Map<String, dynamic> toJson() {
    final json = <String, dynamic>{};
    put(json, 'from', from);
    put(json, 'to', to);
    put(json, 'days', days);
    json['sessions'] = sessionCount;
    json['kinds'] = {
      for (final kind in Kind.values)
        if ((kinds[kind] ?? 0) > 0) kind.name: kinds[kind]!,
    };
    json['rep_bands'] = {
      'sets': {
        for (final band in RepBand.values)
          if ((bandSets[band] ?? 0) > 0) band.name: bandSets[band]!,
      },
      'slots': {
        for (final band in RepBand.values)
          if ((bandSlots[band] ?? 0) > 0) band.name: bandSlots[band]!,
      },
    };
    json['patterns'] = [for (final p in patterns) p.toJson()];
    json['time_domains'] = {
      for (final domain in TimeDomain.values)
        if ((timeDomains[domain] ?? 0) > 0) domain.name: timeDomains[domain]!,
    };
    json['total_tonnage_kg'] = totalTonnageKg;
    json['load'] = [for (final p in load) p.toJson()];
    json['alerts'] = [for (final a in alerts) a.toJson()];
    if (unknownExercises.isNotEmpty) {
      json['unknown_exercises'] = unknownExercises;
    }
    return json;
  }
}

/// The 4 / 2 / 1 shape: three or more recorded sets whose reps only ever fall.
bool _repsCollapse(List<WorkSet> sets) {
  final reps = [
    for (final set in sets)
      if ((set.recordedReps ?? 0) > 0) set.recordedReps!,
  ];
  if (reps.length < 3) return false;
  for (var i = 1; i < reps.length; i++) {
    if (reps[i] >= reps[i - 1]) return false;
  }
  return true;
}

int? _daysBetween(String? from, String? to) {
  if (from == null || to == null) return null;
  final start = DateTime.tryParse(from);
  final end = DateTime.tryParse(to);
  if (start == null || end == null) return null;
  return end.difference(start).inDays + 1;
}

double _round(double value) => (value * 10).round() / 10;

File: crates/workout-log-core/src/analytics/cycle_report.rs
//! What a stretch of sessions says as a block rather than one at a time: band
//! distribution, pattern frequency, combined load, and the alerts those raise.

use std::collections::{BTreeMap, BTreeSet};

use chrono::NaiveDate;
use serde_json::{Map, Value, json};

use crate::analytics::progress::pattern_name;
use crate::analytics::session_metrics::{BandCounts, SessionMetrics, TimeDomain, kind_name};
use crate::models::block::Block;
use crate::models::exercise::{Catalogue, MovementPattern};
use crate::models::session::{Kind, Session};
use crate::models::work_set::{RepBand, WorkSet};

fn round(value: f64) -> f64 {
    (value * 10.0).round() / 10.0
}

/// Dart prints an integral double as `1.0`, and the alert text is part of the
/// contract, so the formatting has to match rather than print `1`.
fn dart_double(value: f64) -> String {
    if value.is_finite() && value.fract() == 0.0 {
        format!("{value:.1}")
    } else {
        format!("{value}")
    }
}

/// A finding worth a training decision, tagged with the principle it comes from.
/// The system reports; the programming decisions stay the athlete's.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TrainingAlert {
    /// `P3`, `P5`, `P8`, `P10` — see `docs/01-training-principles.md`.
    pub principle: String,
    pub message: String,
    pub sessions: Vec<String>,
}

impl TrainingAlert {
    pub fn to_json(&self) -> Value {
        let mut json = Map::new();
        json.insert("principle".into(), json!(self.principle));
        json.insert("message".into(), json!(self.message));
        if !self.sessions.is_empty() {
            json.insert("sessions".into(), json!(self.sessions));
        }
        Value::Object(json)
    }
}

/// How often a movement pattern was trained — the frequency variety is bought
/// with (P8). Below roughly one slot per cycle, linear progression is dead.
#[derive(Debug, Clone, PartialEq)]
pub struct PatternFrequency {
    pub pattern: MovementPattern,
    pub slots: usize,
    pub sessions: usize,
    pub per_cycle: f64,
}

impl PatternFrequency {
    pub fn to_json(&self) -> Value {
        json!({
            "pattern": pattern_name(self.pattern),
            "slots": self.slots,
            "sessions": self.sessions,
            "per_cycle": self.per_cycle,
        })
    }
}

/// Strength and conditioning side by side, never fused into one score: fatigue
/// is additive and the point is to see when both rise together (P7).
#[derive(Debug, Clone, PartialEq)]
pub struct LoadPoint {
    pub date: String,
    pub cycle_day: String,
    pub kind: Kind,
    pub tonnage_kg: f64,
    pub sets: usize,
    pub metcon_sec: Option<f64>,
    pub work_min: Option<f64>,
}

impl LoadPoint {
    pub fn to_json(&self) -> Value {
        let mut json = Map::new();
        json.insert("date".into(), json!(self.date));
        json.insert("cycle_day".into(), json!(self.cycle_day));
        json.insert("kind".into(), json!(kind_name(self.kind)));
        json.insert("tonnage_kg".into(), json!(self.tonnage_kg));
        json.insert("sets".into(), json!(self.sets));
        if let Some(v) = self.metcon_sec {
            json.insert("metcon_sec".into(), json!(v));
        }
        if let Some(v) = self.work_min {
            json.insert("work_min".into(), json!(v));
        }
        Value::Object(json)
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct TrainingReport {
    pub from: Option<String>,
    pub to: Option<String>,
    pub days: Option<i64>,
    pub cycle_length_days: i64,
    pub session_count: usize,
    pub kinds: BTreeMap<Kind, usize>,
    pub band_sets: BandCounts,
    pub band_slots: BandCounts,
    pub patterns: Vec<PatternFrequency>,
    pub load: Vec<LoadPoint>,
    pub time_domains: BTreeMap<String, usize>,
    pub alerts: Vec<TrainingAlert>,
    /// Names no catalogue entry resolves: surfaced, never silently dropped (ADR-006).
    pub unknown_exercises: Vec<String>,
}

const BAND_ORDER: [RepBand; 3] = [RepBand::Heavy, RepBand::Base, RepBand::Volume];

fn band_index(band: RepBand) -> usize {
    match band {
        RepBand::Heavy => 0,
        RepBand::Base => 1,
        RepBand::Volume => 2,
    }
}

fn band_name(band: RepBand) -> &'static str {
    crate::analytics::progress::band_name(band)
}

fn band_json(counts: &BandCounts) -> Value {
    let mut json = Map::new();
    for band in BAND_ORDER {
        let count = counts[band_index(band)];
        if count > 0 {
            json.insert(band_name(band).into(), json!(count));
        }
    }
    Value::Object(json)
}

impl TrainingReport {
    pub fn total_tonnage_kg(&self) -> f64 {
        round(self.load.iter().map(|p| p.tonnage_kg).sum())
    }

    /// `volume_slot_share_threshold` guards P5, the failure that actually
    /// happened: a block with every slot in the volume range left no light
    /// session to recover against and ended in an involuntary deload.
    pub fn of(
        sessions: &[Session],
        catalogue: &Catalogue,
        cycle_length_days: i64,
        volume_slot_share_threshold: f64,
    ) -> Self {
        let mut ordered: Vec<&Session> = sessions.iter().collect();
        ordered.sort_by(|a, b| a.date.cmp(&b.date));

        let mut band_sets: BandCounts = [0; 3];
        let mut band_slots: BandCounts = [0; 3];
        let mut pattern_slots: BTreeMap<MovementPattern, usize> = BTreeMap::new();
        let mut pattern_sessions: BTreeMap<MovementPattern, BTreeSet<String>> = BTreeMap::new();
        let mut time_domains: BTreeMap<String, usize> = BTreeMap::new();
        let mut kinds: BTreeMap<Kind, usize> = BTreeMap::new();
        let mut load = Vec::new();
        let mut unknown: BTreeSet<String> = BTreeSet::new();
        let mut alerts = Vec::new();
        let mut explosive_offenders: BTreeSet<String> = BTreeSet::new();
        let mut decay_offenders: BTreeSet<String> = BTreeSet::new();

        for session in &ordered {
            let metrics = SessionMetrics::of(session);
            *kinds.entry(session.kind).or_insert(0) += 1;
            for index in 0..3 {
                band_sets[index] += metrics.band_sets[index];
                band_slots[index] += metrics.band_slots[index];
            }
            for metcon in &metrics.metcons {
                if let Some(domain) = metcon.domain() {
                    *time_domains.entry(domain.name().to_string()).or_insert(0) += 1;
                }
            }
            load.push(LoadPoint {
                date: session.date.clone(),
                cycle_day: session.cycle_day.clone(),
                kind: session.kind,
                tonnage_kg: metrics.tonnage_kg,
                sets: metrics.sets,
                metcon_sec: (!metrics.metcons.is_empty()).then(|| {
                    metrics
                        .metcons
                        .iter()
                        .map(|m| m.total_sec.unwrap_or(0.0))
                        .sum()
                }),
                work_min: (metrics.density.work_min > 0.0).then_some(metrics.density.work_min),
            });

            for block in &session.blocks {
                let names: Vec<String> = match block {
                    Block::Strength(s) => vec![s.exercise.clone()],
                    Block::Metcon(m) => m.exercises.iter().map(|e| e.name.clone()).collect(),
                    _ => Vec::new(),
                };
                for name in &names {
                    let Some(resolved) = catalogue.resolve(name) else {
                        if !name.trim().is_empty() {
                            unknown.insert(name.clone());
                        }
                        continue;
                    };
                    let Some(pattern) = resolved.pattern else {
                        continue;
                    };
                    if matches!(block, Block::Strength(_)) {
                        *pattern_slots.entry(pattern).or_insert(0) += 1;
                    }
                    pattern_sessions
                        .entry(pattern)
                        .or_default()
                        .insert(session.date.clone());
                }

                if let Block::Strength(strength) = block
                    && let Some(resolved) = catalogue.resolve(&strength.exercise)
                    && resolved.is_explosive()
                {
                    let label = format!("{} {}", session.date, strength.exercise);
                    if strength
                        .sets
                        .iter()
                        .any(|s| s.recorded_reps().unwrap_or(0) > 3)
                    {
                        explosive_offenders.insert(label.clone());
                    }
                    if reps_collapse(&strength.sets) {
                        decay_offenders.insert(label);
                    }
                }
            }
        }

        let from = ordered.first().map(|s| s.date.clone());
        let to = ordered.last().map(|s| s.date.clone());
        let days = days_between(from.as_deref(), to.as_deref());
        let cycles = days.map(|d| d as f64 / cycle_length_days as f64);

        let patterns: Vec<PatternFrequency> = MovementPattern::ALL
            .into_iter()
            .filter(|p| {
                pattern_slots.get(p).copied().unwrap_or(0) > 0
                    || pattern_sessions.get(p).is_some_and(|s| !s.is_empty())
            })
            .map(|pattern| {
                let slots = pattern_slots.get(&pattern).copied().unwrap_or(0);
                PatternFrequency {
                    pattern,
                    slots,
                    sessions: pattern_sessions.get(&pattern).map_or(0, BTreeSet::len),
                    per_cycle: match cycles {
                        Some(c) if c > 0.0 => round(slots as f64 / c),
                        _ => 0.0,
                    },
                }
            })
            .collect();

        let total_slots: usize = band_slots.iter().sum();
        let volume_slots = band_slots[band_index(RepBand::Volume)];
        if total_slots > 0 && volume_slots as f64 / total_slots as f64 > volume_slot_share_threshold
        {
            alerts.push(TrainingAlert {
                principle: "P5".into(),
                message: format!(
                    "volume (7+) slots are {}% of {total_slots} slots, over the {}% guard — the state that produced the involuntary deload",
                    (volume_slots as f64 / total_slots as f64 * 100.0).round(),
                    (volume_slot_share_threshold * 100.0).round()
                ),
                sessions: Vec::new(),
            });
        }
        if !explosive_offenders.is_empty() {
            alerts.push(TrainingAlert {
                principle: "P3".into(),
                message: "explosive lifts logged above 3 reps — bar speed decays and the lift is trained in the wrong zone".into(),
                sessions: explosive_offenders.into_iter().collect(),
            });
        }
        if !decay_offenders.is_empty() {
            alerts.push(TrainingAlert {
                principle: "P3".into(),
                message:
                    "reps collapse across the ramp on an explosive lift (the 4 / 2 / 1 pattern)"
                        .into(),
                sessions: decay_offenders.into_iter().collect(),
            });
        }
        if let Some(cycles) = cycles
            && cycles >= 1.0
        {
            let starved: Vec<String> = patterns
                .iter()
                .filter(|p| p.per_cycle < 1.0)
                .map(|p| {
                    format!(
                        "{} ({}/cycle)",
                        pattern_name(p.pattern),
                        dart_double(p.per_cycle)
                    )
                })
                .collect();
            if !starved.is_empty() {
                alerts.push(TrainingAlert {
                    principle: "P8".into(),
                    message: format!(
                        "trained below once per {cycle_length_days}-day cycle, so linear progression on them is dead: {}",
                        starved.join(", ")
                    ),
                    sessions: Vec::new(),
                });
            }
        }
        if let (Some(cycles), Some(days)) = (cycles, days)
            && cycles >= 2.0
            && kinds.get(&Kind::Deload).copied().unwrap_or(0) == 0
            && kinds.get(&Kind::Retest).copied().unwrap_or(0) == 0
        {
            alerts.push(TrainingAlert {
                principle: "P10".into(),
                message: format!(
                    "no deload or retest in {days} days — deloads are hygiene, and progress between anchors is the unit of evaluation"
                ),
                sessions: Vec::new(),
            });
        }

        Self {
            from,
            to,
            days,
            cycle_length_days,
            session_count: ordered.len(),
            kinds,
            band_sets,
            band_slots,
            patterns,
            load,
            time_domains,
            alerts,
            unknown_exercises: unknown.into_iter().collect(),
        }
    }

    pub fn to_json(&self) -> Value {
        let mut json = Map::new();
        if let Some(from) = &self.from {
            json.insert("from".into(), json!(from));
        }
        if let Some(to) = &self.to {
            json.insert("to".into(), json!(to));
        }
        if let Some(days) = self.days {
            json.insert("days".into(), json!(days));
        }
        json.insert("sessions".into(), json!(self.session_count));
        json.insert(
            "kinds".into(),
            Value::Object(
                Kind::ALL
                    .into_iter()
                    .filter_map(|kind| {
                        let count = self.kinds.get(&kind).copied().unwrap_or(0);
                        (count > 0).then(|| (kind_name(kind).to_string(), json!(count)))
                    })
                    .collect(),
            ),
        );
        json.insert(
            "rep_bands".into(),
            json!({"sets": band_json(&self.band_sets), "slots": band_json(&self.band_slots)}),
        );
        json.insert(
            "patterns".into(),
            Value::Array(
                self.patterns
                    .iter()
                    .map(PatternFrequency::to_json)
                    .collect(),
            ),
        );
        json.insert(
            "time_domains".into(),
            Value::Object(
                TimeDomain::ALL
                    .into_iter()
                    .filter_map(|domain| {
                        let count = self.time_domains.get(domain.name()).copied().unwrap_or(0);
                        (count > 0).then(|| (domain.name().to_string(), json!(count)))
                    })
                    .collect(),
            ),
        );
        json.insert("total_tonnage_kg".into(), json!(self.total_tonnage_kg()));
        json.insert(
            "load".into(),
            Value::Array(self.load.iter().map(LoadPoint::to_json).collect()),
        );
        json.insert(
            "alerts".into(),
            Value::Array(self.alerts.iter().map(TrainingAlert::to_json).collect()),
        );
        if !self.unknown_exercises.is_empty() {
            json.insert("unknown_exercises".into(), json!(self.unknown_exercises));
        }
        Value::Object(json)
    }
}

/// The 4 / 2 / 1 shape: three or more recorded sets whose reps only ever fall.
fn reps_collapse(sets: &[WorkSet]) -> bool {
    let reps: Vec<i64> = sets
        .iter()
        .filter_map(|s| s.recorded_reps().filter(|&r| r > 0))
        .collect();
    reps.len() >= 3 && reps.windows(2).all(|pair| pair[1] < pair[0])
}

fn days_between(from: Option<&str>, to: Option<&str>) -> Option<i64> {
    let parse = |t: &str| NaiveDate::parse_from_str(t, "%Y-%m-%d").ok();
    let (start, end) = (parse(from?)?, parse(to?)?);
    Some((end - start).num_days() + 1)
}
